### DISCUSSION TITLE
tcl-tk is keg-only, could this restriction be removed?

### DISCUSSION THREAD

**Discussion by simonjwright** [Homebrew/discussions#4264] [Everyday usage] [answered]
### Output of `brew config`

```shell
$ brew config
HOMEBREW_VERSION: 4.0.3-96-gd1cba00
ORIGIN: https://github.com/Homebrew/brew
HEAD: d1cba00ee392ba6296f2ca56aadcde5696969d67
Last commit: 2 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 86a596222b5211b28e7b56418f6138031ee9b01c
Core tap last commit: 37 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.5mLe3jb6Ts/org.xquartz:0
HOMEBREW_EDITOR: /Applications/Emacs.app/Contents/MacOS/bin/emacsclient
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.37.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2.1-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
$ brew doctor
Your system is ready to brew.
```

### Description of issue

`brew info tcl-tk` says
```
tcl-tk is keg-only, which means it was not symlinked into /opt/homebrew,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.
```
but if I run `/usr/bin/tclsh` it says
```
WARNING: This version of tcl is included in macOS for compatibility with legacy software. 
In future versions of macOS the tcl runtime will not be available by 
default, and may require you to install an additional package.
```
so (especially bearing in mind that brew happily installs python3, which is also present in `/usr/bin` at version 3.9.6) `tcl-tk`  should be installable normally, not keg-only.

**Comment by gromgit** [comment, marked as answer]
There's another reason why `tcl-tk` is still keg-only, that's currently not possible to work around: https://github.com/Homebrew/homebrew-core/pull/113827#issuecomment-1288372534

In short, anything that depends, directly or indirectly, on both `tcl-tk` and `xorgproto` cannot be built unless `tcl-tk` is keg-only:
```sh
$ brew uses --recursive --eval-all --include-build --include-test --formula tcl-tk xorgproto
arxiv_latex_cleaner       blastem                   liquidctl                 r                         weasyprint
aws-google-auth           fdroidserver              mapproxy                  textract                  woob
bazarr                    gmsh                      ocrmypdf                  torchvision               wxpython
binwalk                   graph-tool                pillow                    trezor-agent
```

**Reply by carlocab** [reply]
We should fix that conflict though. Why is `tcl-tk` shipping X11 headers?

**Reply by gromgit** [reply]
From a quick look at the source, and confirmed by a recursive diff between the `include/X11` dirs of `tcl-tk` and `xorgproto`, it looks like Tk ships modified X11 headers to Aqua-enable Tk-enabled applications, including Tk's own `wish` shell. Rebuilding `tcl-tk` without `--enable-aqua` eliminates the installation of these headers, but reintroduces XQuartz as an unwelcome dependency.

**Reply by carlocab** [reply]
> reintroduces XQuartz as an unwelcome dependency.

Ugh, pass. We should consider updating `tcl-tk`'s keg-only reason, though, instead of it being documented in various discussions/PRs.

**Reply by gromgit** [reply]
Or, as I should've thought of long ago, we could just install all `tcl-tk` headers in a subdirectory: https://github.com/Homebrew/homebrew-core/pull/124056

**Comment by simonjwright** [comment]
Thanks for the info, clearly the restriction can’t be removed any time soon.

I guess this means that "brew link -f tcl-tk" would be a Bad Idea even if brew didn’t reject it outright?

**Reply by gromgit** [reply]
I wouldn't recommend it. Anything you build from then on, that depends on `xorgproto`, would be sourcing Tk's modified X11 headers instead, which likely won't end well.

**Comment by gromgit** [comment]
@simonjwright Problem solved with https://github.com/Homebrew/homebrew-core/pull/124056, `brew update` and `brew upgrade` to get a Tcl/Tk installation that's no longer keg-only.

**Reply by simonjwright** [reply]
Great! on my (intel, Monterey) macbook pro.

However, on Github CI, I said
```
brew update
brew upgrade tcl-tk
```
which failed as in [this log](https://github.com/simonjwright/tcladashell/actions/runs/4263867538/jobs/7421157026#step:5:43) because - I think - they’d already force-linked 8.6.13?? and brew wasn't expecting that?

I will try to unlink before the upgrade.
